framework/Util/Cron/TTimeScheduler.php

```diff
@@ -0,0 +1,856 @@
+					if ($datea['wday'] == 6) {
+						if ($endDay == 1) {
+							$endDay = 3;
+						} else {
+							$endDay--;
+						}
+					} elseif ($datea['wday'] == 0) {
+						if ($endDay == $daysinmonth) {
+							$endDay = $daysinmonth - 2;
+						} else {
+							$endDay++;
+						}
+					}
+				}
+				for ($i = $startDay; $i <= $endDay && $i <= 31; $i += $d['period']) {
+					$da[$i] = 1;
+				}
+			} elseif ($d['dom'] == 'L') {
+				$less = empty($d['end']) ? 0 : $d['end'];
+				$da[$daysinmonth - $less] = 1;
+			}
+		}
+		$firstDatea = getdate(strtotime("$year-$month-01"));
+		foreach ($this->_attr[self::DAY_OF_WEEK] as $d) {
+			if (is_numeric($d['dow'])) {
+				//start at the first sunday on or before the 1st day of the month
+				for ($i = 1 - $firstDatea['wday']; $i <= $daysinmonth; $i += 7) {
+					for ($ii = $d['dow']; ($ii <= $d['end']) && ($ii < 7) && (($i + $ii) <= $daysinmonth); $ii += $d['period']) {
+						$iii = $i + $ii;
+						$w = floor(($iii + 6) / 7);
+						$lw = floor(($daysinmonth - $iii) / 7);
+						
+						if (($iii >= 0) && ((!$d['week'] || $d['week'] == $w) && (!$d['lastDow'] || $lw == 0))) {
+							$dwa[$iii] = 1;
+						}
+					}
+				}
+			}
+		}
+		if ($dowStar) {
+			return $da;
+		} elseif ($domStar) {
+			return $dwa;
+		}
+		foreach ($da as $key => $value) {
+			$da[$key] = $value && ($dwa[$key] ?? 0);
+		}
+		return $da;
+	}
+	
+	
+	/**
+	 * Given which month are valid for the parsed time stamp, this generates
+	 * an array of valid months.  Each month is available as the key to the array
+	 * and the value contains true or false whether the month meets the cron
+	 * time stamp
+	 */
+	protected function getMonthsArray()
+	{
+		$ma = array_pad([], 13, null);
+		unset($ma[0]);
+		foreach ($this->_attr[self::MONTH_OF_YEAR] as $m) {
+			if (is_numeric($m['moy'])) {
+				for ($i = $m['moy']; $i <= $m['end'] && $i <= 12; $i += $m['period']) {
+					$ma[$i] = 1;
+				}
+			}
+		}
+		return $ma;
+	}
+	
+	
+	
+	/**
+	 * Given which years are valid for the parsed time stamp, this generates
+	 * an array of valid years.  Each year is available as the key to the array
+	 * and the value contains true or false whether the year meets the cron
+	 * time stamp.  Only the previos 2 years and next 33 years are available
+	 * @param null|mixed $baseyear
+	 */
+	protected function getYearsArray()
+	{
+		$ya = [];
+		for ($i = self::YEAR_MIN; $i <= self::YEAR_MAX; $i++) {
+			$ya['' . $i] = 0;
+		}
+		foreach ($this->_attr[self::YEAR] as $m) {
+			if (is_numeric($m['year'])) {
+				for ($i = $m['year']; $i <= $m['end']; $i += $m['period']) {
+					$ya[$i] = 1;
+				}
+			}
+		}
+		return $ya;
+	}
+	
+	
+	/**
+	 * This calculates the next trigger time for the schedule based on the parameter
+	 * time.  If no parameter time is given, the current system time is used.
+	 * @param mixed $priortime
+	 * @return numeric the unix time of the next trigger event.
+	 */
+	/*public function getPreviousTriggerTime($priortime = false)
```

Comment:
Is this needed?

Reply:
No.  though if someone in the community wants to write it, for completion, then it would be a gift.  The unit test is huge.